src/coreclr/vm/comdelegate.cpp

```diff
@@ -2189,14 +2188,37 @@ FCIMPL1(PCODE, COMDelegate::GetMulticastInvoke, Object* refThisIn)
         pCode->EmitLabel(nextDelegate);
 
 #ifdef DEBUGGING_SUPPORTED
+        ILCodeLabel *invokeTraceHelper = pCode->NewCodeLabel();
+        ILCodeLabel *debuggerCheckEnd = pCode->NewCodeLabel();
+
+        // Call MulticastDebuggerTraceHelper only if any debugger is attatched
+        pCode->EmitLDC((DWORD_PTR)&g_CORDebuggerControlFlags);
+        pCode->EmitCONV_I();
+        pCode->EmitLDIND_I4();
+
+        // (g_CORDebuggerControlFlags & DBCF_ATTACHED) != 0
+        pCode->EmitLDC(DBCF_ATTACHED);
+        pCode->EmitAND();
+        pCode->EmitBRTRUE(invokeTraceHelper);
+        pCode->EmitBR(debuggerCheckEnd); // Tune branch prediction to prefer non-debugging path
+
+        pCode->EmitLabel(invokeTraceHelper);
+
         pCode->EmitLoadThis();
         pCode->EmitLDLOC(dwLoopCounterNum);
         pCode->EmitCALL(METHOD__STUBHELPERS__MULTICAST_DEBUGGER_TRACE_HELPER, 2, 0);
```

Comment:
```suggestion
        pCode->EmitBRTRUE(invokeTraceHelper);
```

And move the debugging path to be after `RET`:

```
        pCode->EmitRET();

#ifdef DEBUGGING_SUPPORTED
        // Emit debugging support at the end of the method for better perf
        pCode->EmitLabel(invokeTraceHelper);

        pCode->EmitLoadThis();
        pCode->EmitLDLOC(dwLoopCounterNum);
        pCode->EmitCALL(METHOD__STUBHELPERS__MULTICAST_DEBUGGER_TRACE_HELPER, 2, 0);

        pCode->EmitBR(debuggerCheckEnd);
#endif
```

This should be even better

Reply:
That was exactly what I did locally, and resulted in exact same codegen with current, at least for x86.

@dotnet/jit-contrib Do you have any suggestion on this?

Reply:
Well normalizing the loop closer to a `for` loop results in about 5% improvement for ManyArg case.

Reply:
Codegen for x64 now:

```asm
; Assembly listing for method System.Action:IL_STUB_MulticastDelegate_Invoke():this (FullOpts)
; Emitting BLENDED_CODE for X64 with AVX - Windows
; FullOpts code
; optimized code
; optimized using Synthesized PGO
; rsp based frame
; partially interruptible
; with Synthesized PGO: fgCalledCount is 100
; No PGO data

G_M000_IG01:                ;; offset=0x0000
       push     rsi
       push     rbx
       sub      rsp, 40
       mov      rbx, rcx

G_M000_IG02:                ;; offset=0x0009
       xor      esi, esi
       cmp      qword ptr [rbx+0x30], 0
       jle      SHORT G_M000_IG05

G_M000_IG03:                ;; offset=0x0012
       test     dword ptr [(reloc 0x7ff824c55210)], 512
       jne      SHORT G_M000_IG06

G_M000_IG04:                ;; offset=0x001E
       mov      rcx, gword ptr [rbx+0x28]
       cmp      esi, dword ptr [rcx+0x08]
       jae      SHORT G_M000_IG07
       mov      rax, gword ptr [rcx+8*rsi+0x10]
       mov      rcx, gword ptr [rax+0x08]
       call     [rax+0x18]System.Action:Invoke():this
       inc      esi
       movsxd   rcx, esi
       cmp      rcx, qword ptr [rbx+0x30]
       jl       SHORT G_M000_IG03

G_M000_IG05:                ;; offset=0x003E
       add      rsp, 40
       pop      rbx
       pop      rsi
       ret

G_M000_IG06:                ;; offset=0x0045
       mov      rcx, rbx
       mov      edx, esi
       call     System.StubHelpers.StubHelpers:MulticastDebuggerTraceHelper(System.Object,int)
       jmp      SHORT G_M000_IG04

G_M000_IG07:                ;; offset=0x0051
       call     CORINFO_HELP_RNGCHKFAIL
       int3

; Total bytes of code 87
```